### ISSUE TITLE
Sql repository

### ISSUE THREAD

**Issue by pawelszymczyk** [javers/javers#67]
As a relational database user i want to persists changes in Sql reposioty.

**Comment by bartoszwalacik** [comment]
what about read ? :)

**Comment by pawelszymczyk** [comment]
Still some problems with H2 in mem

**Comment by pawelszymczyk** [comment]
def maxSnapshotPk = SELECT MAX(snapshot_pk) FROM jv_snapshot, jv_global_id, jv_cdo_class WHERE qualified_name = 'org.javers.core.model.SnapshotEntity'  LIMIT 2 OFFSET 0

SELECT type FROM "public"."jv_snapshot" WHERE "public"."jv_snapshot"."snapshot_pk" = maxSnapshotPk

SELECT "jv_commit"."author", "jv_commit"."commit_date", "jv_commit"."commit_id" FROM "public"."jv_commit", "public"."jv_snapshot" where "jv_snapshot"."snapshot_pk" = maxSnapshotPk

SELECT "public"."jv_snapshot_property"."name", "public"."jv_snapshot_property"."value" FROM "public"."jv_snapshot_property" where "public"."jv_snapshot_property"."snapshot_fk" = $fk

smth like this ^^

**Comment by pawelszymczyk** [comment]
btw MAX works fine, it was my mistake ;)

**Comment by bartoszwalacik** [comment]
- trzeba usunąc limit z zapytań z MAX
- dodać joiny :)

może 
Snapshot snapshotRepository getById(int snapshoId)
?

i wtedy Snapshot powinien mieć zbudowany commit i propertisy

**Comment by efreet007** [comment]
Hi
When do you plan to implement this feature? Will it work with MS SQL Server?

**Comment by pawelszymczyk** [comment]
Hi
We use https://github.com/polyjdbc/polyjdbc as rapper under standard JDBC driver, in current version Poly supperts only:
* H2
* PostgreSQL
* MySQL
* Oracle (without limit/offset in SELECT)

if you want to use MS SQL Server it should be implemented in Poly.

**Comment by bartoszwalacik** [comment]
Pawel wanted to say, that we add MS SQL support to PolyJDBC after we finished 
work with this issue. Now we are focused on open databases. Adding support for commercial databases (Oracle, MS SQL) is the next step. Rough estimate for this is 2015.03.

efreet007, would you like to contribute to JaVars and help us with SQL stuff?

**Comment by bartoszwalacik** [comment]
efreet007  please put new issue for MS SQL

**Comment by efreet007** [comment]
Done: polyjdbc/polyjdbc#9
Regarding contribution, I do not promise but I will take a look around

**Comment by bartoszwalacik** [comment]
efreet007
I've created separeted JaVeres issue for that
https://github.com/javers/javers/issues/100
please put a comment there to receive notifications

**Comment by bartoszwalacik** [comment]
@efreet007
JaversSqlRepository beta is available for testing
see version 1.0.7-SNAPSHOT
in https://oss.sonatype.org/content/repositories/snapshots/
for now there is support for H2, MySql and Postgres

**Comment by bartoszwalacik** [comment]
config example
```groovy
        def dbConnection = DriverManager.getConnection("jdbc:h2:tcp://localhost:9092/mem:test")
        def connectionProvider = new ConnectionProvider() {
            @Override
            Connection getConnection() {
               return dbConnection
            }
        }
        
        def sqlRepository = SqlRepositoryBuilder
                .sqlRepository()
                .withConnectionProvider(connectionProvider)
                .withDialect(DialectName.H2).build()
        javers = javers().registerJaversRepository(sqlRepository).build()
        sqlRepository.setJsonConverter(javers.jsonConverter)
```

**Comment by igler** [comment]
Is it possible to use Hibernate Envers in combination with JaVers? Envers can track changes though annotating the classes with

<code>@Audited(withModifiedFlag = true)</code>

 The idea is to pull the changes from the _AUD tables?

**Comment by bartoszwalacik** [comment]
@igler so envers cant calculate diffs on its own?

**Comment by igler** [comment]
Unfortunately not. You only get a list of revisions by spring-data-envers. So you have to find out the deltas on your own. The idea was if I can submit a list of revisions to JaVers and get a nice JSON-history back?

**Comment by bartoszwalacik** [comment]
@igler maybe we move this new envers thread to a new issue, titled 'javers envers integration'?

Your idea sounds good, we can think about read-only JaversRepository implementation (see https://github.com/javers/javers/blob/033fe77817052e4ec71b9014f676c5bfa478cbc8/javers-core/src/main/java/org/javers/repository/api/JaversRepository.java), which would read object snapshots from envers tables.
Are you eager to contribute to development of this new feature?

**Comment by igler** [comment]
OK, moved the discussion to https://github.com/javers/javers/issues/112.

**Comment by bartoszwalacik** [comment]
@efreet007, @igler, JaversSqlRepository is done and released as 1.1.0-RC1.
It would be nice if you test it and give us some feedback before final 1.1.0 release

**Comment by bartoszwalacik** [comment]
released in v. 1.1.0
